packages/angular_devkit/build_angular/src/protractor/index2.ts

```diff
@@ -0,0 +1,148 @@
+  }
+
+  if (!path) {
+    throw new Error(tags.stripIndents`
+      Cannot automatically find webdriver-manager to update.
+      Update webdriver-manager manually and run 'ng e2e --no-webdriver-update' instead.
+    `);
+  }
+
+  // tslint:disable-next-line:max-line-length no-implicit-dependencies
+  const webdriverUpdate = await import(path) as typeof import ('webdriver-manager/built/lib/cmds/update');
+
+  // run `webdriver-manager update --standalone false --gecko false --quiet`
+  // if you change this, update the command comment in prev line
+  return webdriverUpdate.program.run({
+    standalone: false,
+    gecko: false,
+    quiet: true,
+  } as unknown as JSON);
+}
+
+export async function execute(
+  options: ProtractorBuilderOptions,
+  context: BuilderContext,
+): Promise<BuilderOutput> {
+  // ensure that only one of these options is used
+  if (options.devServerTarget && options.baseUrl) {
+    throw new Error(tags.stripIndents`
+    The 'baseUrl' option cannot be used with 'devServerTarget'.
+    When present, 'devServerTarget' will be used to automatically setup 'baseUrl' for Protractor.
+    `);
+  }
+
+  if (options.webdriverUpdate) {
+    try {
+      await updateWebdriver();
+    } catch (error) {
+      context.reportStatus('Error: ' + error);
+
+      return { success: false };
+    }
+  }
+
+  let baseUrl;
+  let server;
+  if (options.devServerTarget) {
+    const parts = (options.devServerTarget as string).split(':');
+    const target = { project: parts[0], target: parts[1], configuration: parts[2] };
+    const serverOptions = await context.getTargetOptions(target);
+
+    const overrides: Record<string, string | number | boolean > = { watch: false };
+    if (options.host !== undefined) { overrides.host = options.host; }
+    if (options.port !== undefined) { overrides.port = options.port; }
+    server = await context.scheduleTarget(target, overrides);
+
+    let result;
+    try {
+      result = await server.result;
+    } catch (error) {
+
+    }
+
+    if (!result || !result.success) {
+      return { success: false };
+    }
+
+    if (typeof serverOptions.publicHost === 'string') {
+      let publicHost = serverOptions.publicHost as string;
+      if (!/^\w+:\/\//.test(publicHost)) {
+        publicHost = `${serverOptions.ssl
+          ? 'https'
+          : 'http'}://${publicHost}`;
+      }
+      const clientUrl = url.parse(publicHost);
+      baseUrl = url.format(clientUrl);
+    } else if (typeof result.port === 'number') {
+      baseUrl = url.format({
+        protocol: serverOptions.ssl ? 'https' : 'http',
+        hostname: options.host,
+        port: result.port.toString(),
+      });
+    }
+  }
+
+  try {
+    return await runProtractor(context.workspaceRoot, { ...options, baseUrl });
+  } catch {
+    return { success: false };
+  } finally {
+    if (server) {
+      await server.stop();
+    }
+  }
+}
+
+export default createBuilder<Record<string, string> & ProtractorBuilderOptions>(execute);
```

Comment:
What's the rationale of using `Record<string, string>` over `JsonObject`?

Reply:
It made the compiler stop complaining and didn't require an extra import.  `JsonObject` is functionally a `Record<string, string | boolean | ...>` and the compiler just wanted something with an index signature.

Using either option breaks the type constraint, the following will happily compile for instance:
```
export default createBuilder<JsonObject & ProtractorBuilderOptions & { f: Function }>(execute);
```

It may be better to just drop the constraint.  A more comprehensive runtime check could be added if needed.